File: clients/nodejs/src/protocol.ts
/**
 * SixSevenDB wire protocol v3 message encoding and decoding.
 *
 * The wire protocol follows PostgreSQL v3 framing:
 *   Frontend (client → server): 1-byte type + 4-byte length + payload
 *   Backend  (server → client): 1-byte type + 4-byte length + payload
 *   Startup message has no type byte: 4-byte length + payload
 */

import { createHash, createHmac, pbkdf2Sync, randomBytes } from 'crypto';

// ---------------------------------------------------------------------------
// Frontend message builders
// ---------------------------------------------------------------------------

export function buildStartupMessage(user: string, database: string): Buffer {
  const PROTOCOL_VERSION = 196608; // 3 << 16
  const params = Buffer.from(`user\0${user}\0database\0${database}\0\0`, 'utf8');
  const len = 4 + 4 + params.length; // length field + version + params
  const buf = Buffer.alloc(len);
  buf.writeInt32BE(len, 0);
  buf.writeInt32BE(PROTOCOL_VERSION, 4);
  params.copy(buf, 8);
  return buf;
}

export function buildPasswordMessage(password: string): Buffer {
  const payload = Buffer.from(password + '\0', 'utf8');
  const len = 4 + payload.length;
  const buf = Buffer.alloc(1 + len);
  buf[0] = 0x70; // 'p'
  buf.writeInt32BE(len, 1);
  payload.copy(buf, 5);
  return buf;
}

export function buildMD5PasswordMessage(
  user: string,
  password: string,
  salt: Buffer,
): Buffer {
  // md5(md5(password + user) + salt)
  const inner = createHash('md5')
    .update(password + user)
    .digest('hex');
  const outer = createHash('md5')
    .update(Buffer.concat([Buffer.from(inner, 'utf8'), salt]))
    .digest('hex');
  return buildPasswordMessage('md5' + outer);
}

export function buildQueryMessage(sql: string): Buffer {
  const payload = Buffer.from(sql + '\0', 'utf8');
  const len = 4 + payload.length;
  const buf = Buffer.alloc(1 + len);
  buf[0] = 0x51; // 'Q'
  buf.writeInt32BE(len, 1);
  payload.copy(buf, 5);
  return buf;
}

export function buildParseMessage(
  sql: string,
  statementName = '',
): Buffer {
  const nameBytes = Buffer.from(statementName + '\0', 'utf8');
  const queryBytes = Buffer.from(sql + '\0', 'utf8');
  const len = 4 + nameBytes.length + queryBytes.length + 2; // +2 for int16 param count
  const buf = Buffer.alloc(1 + len);
  let offset = 0;
  buf[offset++] = 0x50; // 'P'
  buf.writeInt32BE(len, offset);
  offset += 4;
  nameBytes.copy(buf, offset);
  offset += nameBytes.length;
  queryBytes.copy(buf, offset);
  offset += queryBytes.length;
  buf.writeInt16BE(0, offset); // no param type OIDs
  return buf;
}

export function buildBindMessage(
  params: (string | null)[],
  portalName = '',
  statementName = '',
): Buffer {
  const portalBytes = Buffer.from(portalName + '\0', 'utf8');
  const stmtBytes = Buffer.from(statementName + '\0', 'utf8');

  // Compute total size
  let payloadSize =
    portalBytes.length +
    stmtBytes.length +
    2 + // int16: number of format codes
    2 + // int16: number of parameters
    2;  // int16: number of result format codes

  const paramBuffers: (Buffer | null)[] = [];
  for (const p of params) {
    if (p === null) {
      paramBuffers.push(null);
      payloadSize += 4; // int32 -1
    } else {
      const b = Buffer.from(p, 'utf8');
      paramBuffers.push(b);
      payloadSize += 4 + b.length; // int32 length + data
    }
  }

  const len = 4 + payloadSize;
  const buf = Buffer.alloc(1 + len);
  let offset = 0;
  buf[offset++] = 0x42; // 'B'
  buf.writeInt32BE(len, offset);
  offset += 4;
  portalBytes.copy(buf, offset);
  offset += portalBytes.length;
  stmtBytes.copy(buf, offset);
  offset += stmtBytes.length;
  // Format codes: 0 = all text
  buf.writeInt16BE(0, offset);
  offset += 2;
  // Number of parameters
  buf.writeInt16BE(params.length, offset);
  offset += 2;
  // Parameter values
  for (const pb of paramBuffers) {
    if (pb === null) {
      buf.writeInt32BE(-1, offset);
      offset += 4;
    } else {
      buf.writeInt32BE(pb.length, offset);
      offset += 4;
      pb.copy(buf, offset);
      offset += pb.length;
    }
  }
  // Result format codes: 0 = all text
  buf.writeInt16BE(0, offset);
  return buf;
}

export function buildDescribeMessage(type: 'S' | 'P' = 'P', name = ''): Buffer {
  const nameBytes = Buffer.from(name + '\0', 'utf8');
  const len = 4 + 1 + nameBytes.length;
  const buf = Buffer.alloc(1 + len);
  buf[0] = 0x44; // 'D'
  buf.writeInt32BE(len, 1);
  buf[5] = type.charCodeAt(0);
  nameBytes.copy(buf, 6);
  return buf;
}

export function buildExecuteMessage(portalName = '', maxRows = 0): Buffer {
  const nameBytes = Buffer.from(portalName + '\0', 'utf8');
  const len = 4 + nameBytes.length + 4;
  const buf = Buffer.alloc(1 + len);
  buf[0] = 0x45; // 'E'
  buf.writeInt32BE(len, 1);
  nameBytes.copy(buf, 5);
  buf.writeInt32BE(maxRows, 5 + nameBytes.length);
  return buf;
}

export function buildSyncMessage(): Buffer {
  const buf = Buffer.alloc(5);
  buf[0] = 0x53; // 'S'
  buf.writeInt32BE(4, 1);
  return buf;
}

export function buildTerminateMessage(): Buffer {
  const buf = Buffer.alloc(5);
  buf[0] = 0x58; // 'X'
  buf.writeInt32BE(4, 1);
  return buf;
}

export function buildSASLInitialResponse(mechanism: string, clientFirstMessage: string): Buffer {
  const mechBytes = Buffer.from(mechanism + '\0', 'utf8');
  const msgBytes = Buffer.from(clientFirstMessage, 'utf8');
  const len = 4 + mechBytes.length + 4 + msgBytes.length;
  const buf = Buffer.alloc(1 + len);
  let offset = 0;
  buf[offset++] = 0x70; // 'p'
  buf.writeInt32BE(len, offset);
  offset += 4;
  mechBytes.copy(buf, offset);
  offset += mechBytes.length;
  buf.writeInt32BE(msgBytes.length, offset);
  offset += 4;
  msgBytes.copy(buf, offset);
  return buf;
}

export function buildSASLResponse(clientFinalMessage: string): Buffer {
  const msgBytes = Buffer.from(clientFinalMessage, 'utf8');
  const len = 4 + msgBytes.length;
  const buf = Buffer.alloc(1 + len);
  buf[0] = 0x70; // 'p'
  buf.writeInt32BE(len, 1);
  msgBytes.copy(buf, 5);
  return buf;
}

export function buildCloseMessage(type: 'S' | 'P', name = ''): Buffer {
  const nameBytes = Buffer.from(name + '\0', 'utf8');
  const len = 4 + 1 + nameBytes.length;
  const buf = Buffer.alloc(1 + len);
  buf[0] = 0x43; // 'C'
  buf.writeInt32BE(len, 1);
  buf[5] = type.charCodeAt(0);
  nameBytes.copy(buf, 6);
  return buf;
}

// ---------------------------------------------------------------------------
// Backend message types
// ---------------------------------------------------------------------------

export const enum BackendMessageType {
  AuthenticationOk = 'AuthenticationOk',
  AuthenticationCleartextPassword = 'AuthenticationCleartextPassword',
  AuthenticationMD5Password = 'AuthenticationMD5Password',
  AuthenticationSASL = 'AuthenticationSASL',
  AuthenticationSASLContinue = 'AuthenticationSASLContinue',
  AuthenticationSASLFinal = 'AuthenticationSASLFinal',
  ParameterStatus = 'ParameterStatus',
  BackendKeyData = 'BackendKeyData',
  ReadyForQuery = 'ReadyForQuery',
  RowDescription = 'RowDescription',
  DataRow = 'DataRow',
  CommandComplete = 'CommandComplete',
  ErrorResponse = 'ErrorResponse',
  NoticeResponse = 'NoticeResponse',
  ParseComplete = 'ParseComplete',
  BindComplete = 'BindComplete',
  CloseComplete = 'CloseComplete',
  NoData = 'NoData',
  EmptyQueryResponse = 'EmptyQueryResponse',
}

export interface FieldDescription {
  name: string;
  tableOID: number;
  columnIndex: number;
  typeOID: number;
  typeSize: number;
  typeModifier: number;
  formatCode: number;
}

export type BackendMessage =
  | { type: BackendMessageType.AuthenticationOk }
  | { type: BackendMessageType.AuthenticationCleartextPassword }
  | { type: BackendMessageType.AuthenticationMD5Password; salt: Buffer }
  | { type: BackendMessageType.AuthenticationSASL; mechanisms: string[] }
  | { type: BackendMessageType.AuthenticationSASLContinue; data: string }
  | { type: BackendMessageType.AuthenticationSASLFinal; data: string }
  | { type: BackendMessageType.ParameterStatus; name: string; value: string }
  | { type: BackendMessageType.BackendKeyData; processId: number; secretKey: number }
  | { type: BackendMessageType.ReadyForQuery; status: string }
  | { type: BackendMessageType.RowDescription; fields: FieldDescription[] }
  | { type: BackendMessageType.DataRow; values: (string | null)[] }
  | { type: BackendMessageType.CommandComplete; tag: string }
  | { type: BackendMessageType.ErrorResponse; severity: string; code: string; message: string }
  | { type: BackendMessageType.NoticeResponse; severity: string; message: string }
  | { type: BackendMessageType.ParseComplete }
  | { type: BackendMessageType.BindComplete }
  | { type: BackendMessageType.CloseComplete }
  | { type: BackendMessageType.NoData }
  | { type: BackendMessageType.EmptyQueryResponse };

// ---------------------------------------------------------------------------
// Backend message parsing
// ---------------------------------------------------------------------------

function readCString(buf: Buffer, offset: number): { value: string; end: number } {
  const nullPos = buf.indexOf(0, offset);
  if (nullPos === -1) throw new Error('unterminated string in message');
  return { value: buf.toString('utf8', offset, nullPos), end: nullPos + 1 };
}

function parseBackendMessage(typeByte: number, payload: Buffer): BackendMessage {
  switch (typeByte) {
    case 0x52: { // 'R' Authentication
      const authType = payload.readInt32BE(0);
      if (authType === 0) return { type: BackendMessageType.AuthenticationOk };
      if (authType === 3) return { type: BackendMessageType.AuthenticationCleartextPassword };
      if (authType === 5) {
        return {
          type: BackendMessageType.AuthenticationMD5Password,
          salt: Buffer.from(payload.subarray(4, 8)),
        };
      }
      if (authType === 10) {
        // AuthenticationSASL — list of mechanisms
        const mechanisms: string[] = [];
        let offset = 4;
        while (offset < payload.length) {
          const { value, end } = readCString(payload, offset);
          if (value === '') break;
          mechanisms.push(value);
          offset = end;
        }
        return { type: BackendMessageType.AuthenticationSASL, mechanisms };
      }
      if (authType === 11) {
        return {
          type: BackendMessageType.AuthenticationSASLContinue,
          data: payload.toString('utf8', 4),
        };
      }
      if (authType === 12) {
        return {
          type: BackendMessageType.AuthenticationSASLFinal,
          data: payload.toString('utf8', 4),
        };
      }
      throw new Error(`unsupported authentication type: ${authType}`);
    }

    case 0x53: { // 'S' ParameterStatus
      const { value: name, end: e1 } = readCString(payload, 0);
      const { value } = readCString(payload, e1);
      return { type: BackendMessageType.ParameterStatus, name, value };
    }

    case 0x4b: { // 'K' BackendKeyData
      return {
        type: BackendMessageType.BackendKeyData,
        processId: payload.readInt32BE(0),
        secretKey: payload.readInt32BE(4),
      };
    }

    case 0x5a: { // 'Z' ReadyForQuery
      return {
        type: BackendMessageType.ReadyForQuery,
        status: String.fromCharCode(payload[0]),
      };
    }

    case 0x54: { // 'T' RowDescription
      const fieldCount = payload.readInt16BE(0);
      let offset = 2;
      const fields: FieldDescription[] = [];
      for (let i = 0; i < fieldCount; i++) {
        const { value: name, end } = readCString(payload, offset);
        offset = end;
        const tableOID = payload.readInt32BE(offset); offset += 4;
        const columnIndex = payload.readInt16BE(offset); offset += 2;
        const typeOID = payload.readInt32BE(offset); offset += 4;
        const typeSize = payload.readInt16BE(offset); offset += 2;
        const typeModifier = payload.readInt32BE(offset); offset += 4;
        const formatCode = payload.readInt16BE(offset); offset += 2;
        fields.push({ name, tableOID, columnIndex, typeOID, typeSize, typeModifier, formatCode });
      }
      return { type: BackendMessageType.RowDescription, fields };
    }

    case 0x44: { // 'D' DataRow
      const colCount = payload.readInt16BE(0);
      let offset = 2;
      const values: (string | null)[] = [];
      for (let i = 0; i < colCount; i++) {
        const len = payload.readInt32BE(offset);
        offset += 4;
        if (len === -1) {
          values.push(null);
        } else {
          values.push(payload.toString('utf8', offset, offset + len));
          offset += len;
        }
      }
      return { type: BackendMessageType.DataRow, values };
    }

    case 0x43: { // 'C' CommandComplete
      const { value: tag } = readCString(payload, 0);
      return { type: BackendMessageType.CommandComplete, tag };
    }

    case 0x45: { // 'E' ErrorResponse
      let severity = '';
      let code = '';
      let message = '';
      let offset = 0;
      while (offset < payload.length) {
        const fieldType = payload[offset++];
        if (fieldType === 0) break;
        const { value, end } = readCString(payload, offset);
        offset = end;
        if (fieldType === 0x53) severity = value;     // 'S'
        else if (fieldType === 0x43) code = value;     // 'C'
        else if (fieldType === 0x4d) message = value;  // 'M'
      }
      return { type: BackendMessageType.ErrorResponse, severity, code, message };
    }

    case 0x4e: { // 'N' NoticeResponse
      let severity = '';
      let message = '';
      let offset = 0;
      while (offset < payload.length) {
        const fieldType = payload[offset++];
        if (fieldType === 0) break;
        const { value, end } = readCString(payload, offset);
        offset = end;
        if (fieldType === 0x53) severity = value;
        else if (fieldType === 0x4d) message = value;
      }
      return { type: BackendMessageType.NoticeResponse, severity, message };
    }

    case 0x31: return { type: BackendMessageType.ParseComplete };  // '1'
    case 0x32: return { type: BackendMessageType.BindComplete };   // '2'
    case 0x33: return { type: BackendMessageType.CloseComplete };  // '3'
    case 0x6e: return { type: BackendMessageType.NoData };         // 'n'
    case 0x49: return { type: BackendMessageType.EmptyQueryResponse }; // 'I'

    default:
      throw new Error(`unknown backend message type: 0x${typeByte.toString(16)}`);
  }
}

// ---------------------------------------------------------------------------
// Message stream reader — accumulates TCP data and yields complete messages
// ---------------------------------------------------------------------------

export class MessageReader {
  private buffer: Buffer = Buffer.alloc(0);

  /** Append raw data received from the socket. */
  append(data: Buffer): void {
    this.buffer = Buffer.concat([this.buffer, data]);
  }

  /** Try to read one complete backend message. Returns null if not enough data. */
  read(): BackendMessage | null {
    if (this.buffer.length < 5) return null;

    const typeByte = this.buffer[0];
    const bodyLength = this.buffer.readInt32BE(1); // includes self (4 bytes), excludes type byte
    const totalLength = 1 + bodyLength;

    if (this.buffer.length < totalLength) return null;

    const payload = this.buffer.subarray(5, totalLength);
    this.buffer = this.buffer.subarray(totalLength);

    return parseBackendMessage(typeByte, payload);
  }
}

// ---------------------------------------------------------------------------
// SCRAM-SHA-256 helpers (RFC 5802)
// ---------------------------------------------------------------------------

export function generateClientNonce(length = 24): string {
  return randomBytes(length).toString('base64');
}

export function buildClientFirstMessage(user: string, nonce: string): string {
  // n,,n=<user>,r=<nonce>
  // gs2-header = "n,," (no channel binding)
  return `n,,n=${saslPrepName(user)},r=${nonce}`;
}

export function parseServerFirstMessage(data: string): {
  nonce: string;
  salt: Buffer;
  iterations: number;
} {
  const parts = new Map<string, string>();
  for (const attr of data.split(',')) {
    const eq = attr.indexOf('=');
    if (eq > 0) parts.set(attr[0], attr.substring(eq + 1));
  }
  const nonce = parts.get('r');
  const saltB64 = parts.get('s');
  const iterStr = parts.get('i');
  if (!nonce || !saltB64 || !iterStr) {
    throw new Error('invalid SCRAM server-first-message');
  }
  return {
    nonce,
    salt: Buffer.from(saltB64, 'base64'),
    iterations: parseInt(iterStr, 10),
  };
}

export function computeSaltedPassword(password: string, salt: Buffer, iterations: number): Buffer {
  return pbkdf2Sync(password, salt, iterations, 32, 'sha256');
}

export function computeClientProof(
  saltedPassword: Buffer,
  authMessage: string,
): { clientProof: string; serverSignature: string } {
  const clientKey = hmacSha256(saltedPassword, 'Client Key');
  const storedKey = createHash('sha256').update(clientKey).digest();
  const clientSignature = hmacSha256(storedKey, authMessage);
  const clientProof = Buffer.alloc(clientKey.length);
  for (let i = 0; i < clientKey.length; i++) {
    clientProof[i] = clientKey[i] ^ clientSignature[i];
  }

  const serverKey = hmacSha256(saltedPassword, 'Server Key');
  const serverSignature = hmacSha256(serverKey, authMessage);

  return {
    clientProof: clientProof.toString('base64'),
    serverSignature: serverSignature.toString('base64'),
  };
}

export function buildClientFinalMessage(
  clientFirstBare: string,
  serverFirstMessage: string,
  serverNonce: string,
  clientProof: string,
): string {
  const channelBinding = Buffer.from('n,,').toString('base64');
  const clientFinalWithoutProof = `c=${channelBinding},r=${serverNonce}`;
  return `${clientFinalWithoutProof},p=${clientProof}`;
}

export function buildAuthMessage(
  clientFirstBare: string,
  serverFirstMessage: string,
  serverNonce: string,
): string {
  const channelBinding = Buffer.from('n,,').toString('base64');
  const clientFinalWithoutProof = `c=${channelBinding},r=${serverNonce}`;
  return `${clientFirstBare},${serverFirstMessage},${clientFinalWithoutProof}`;
}

export function verifyServerSignature(serverFinalMessage: string, expectedSignature: string): void {
  const parts = new Map<string, string>();
  for (const attr of serverFinalMessage.split(',')) {
    const eq = attr.indexOf('=');
    if (eq > 0) parts.set(attr[0], attr.substring(eq + 1));
  }
  const v = parts.get('v');
  if (v !== expectedSignature) {
    throw new Error('SCRAM server signature verification failed');
  }
}

function hmacSha256(key: Buffer, data: string): Buffer {
  return createHmac('sha256', key).update(data).digest();
}

function saslPrepName(name: string): string {
  return name.replace(/=/g, '=3D').replace(/,/g, '=2C');
}

File: clients/python/src/giodb/protocol.py
"""PostgreSQL v3 wire protocol implementation for GioDB.

Implements frontend (client -> server) message builders and backend
(server -> client) message parsing compatible with the PG wire protocol.
"""

from __future__ import annotations

import hashlib
import struct
from enum import IntEnum
from typing import Any


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

PROTOCOL_VERSION = 196608  # 3 << 16 (PG v3)


class BackendMessageType(IntEnum):
    AUTHENTICATION = ord("R")
    PARAMETER_STATUS = ord("S")
    BACKEND_KEY_DATA = ord("K")
    READY_FOR_QUERY = ord("Z")
    ROW_DESCRIPTION = ord("T")
    DATA_ROW = ord("D")
    COMMAND_COMPLETE = ord("C")
    ERROR_RESPONSE = ord("E")
    NOTICE_RESPONSE = ord("N")
    EMPTY_QUERY_RESPONSE = ord("I")
    PARSE_COMPLETE = ord("1")
    BIND_COMPLETE = ord("2")
    NO_DATA = ord("n")


# ---------------------------------------------------------------------------
# Backend message structures
# ---------------------------------------------------------------------------


class BackendMessage:
    """Base class for parsed backend messages."""

    msg_type: int


class AuthenticationOk(BackendMessage):
    msg_type = BackendMessageType.AUTHENTICATION
    auth_type = 0


class AuthenticationCleartextPassword(BackendMessage):
    msg_type = BackendMessageType.AUTHENTICATION
    auth_type = 3


class AuthenticationMD5Password(BackendMessage):
    msg_type = BackendMessageType.AUTHENTICATION
    auth_type = 5

    def __init__(self, salt: bytes) -> None:
        self.salt = salt


class AuthenticationSASL(BackendMessage):
    msg_type = BackendMessageType.AUTHENTICATION
    auth_type = 10

    def __init__(self, mechanisms: list[str]) -> None:
        self.mechanisms = mechanisms


class AuthenticationSASLContinue(BackendMessage):
    msg_type = BackendMessageType.AUTHENTICATION
    auth_type = 11

    def __init__(self, data: bytes) -> None:
        self.data = data


class AuthenticationSASLFinal(BackendMessage):
    msg_type = BackendMessageType.AUTHENTICATION
    auth_type = 12

    def __init__(self, data: bytes) -> None:
        self.data = data


class ParameterStatus(BackendMessage):
    msg_type = BackendMessageType.PARAMETER_STATUS

    def __init__(self, name: str, value: str) -> None:
        self.name = name
        self.value = value


class BackendKeyData(BackendMessage):
    msg_type = BackendMessageType.BACKEND_KEY_DATA

    def __init__(self, process_id: int, secret_key: int) -> None:
        self.process_id = process_id
        self.secret_key = secret_key


class ReadyForQuery(BackendMessage):
    msg_type = BackendMessageType.READY_FOR_QUERY

    def __init__(self, status: str) -> None:
        self.status = status  # 'I' idle, 'T' transaction, 'E' error


class FieldDescription:
    def __init__(
        self,
        name: str,
        table_oid: int,
        column_index: int,
        type_oid: int,
        type_size: int,
        type_modifier: int,
        format_code: int,
    ) -> None:
        self.name = name
        self.table_oid = table_oid
        self.column_index = column_index
        self.type_oid = type_oid
        self.type_size = type_size
        self.type_modifier = type_modifier
        self.format_code = format_code


class RowDescription(BackendMessage):
    msg_type = BackendMessageType.ROW_DESCRIPTION

    def __init__(self, fields: list[FieldDescription]) -> None:
        self.fields = fields


class DataRow(BackendMessage):
    msg_type = BackendMessageType.DATA_ROW

    def __init__(self, values: list[bytes | None]) -> None:
        self.values = values


class CommandComplete(BackendMessage):
    msg_type = BackendMessageType.COMMAND_COMPLETE

    def __init__(self, tag: str) -> None:
        self.tag = tag


class ErrorResponse(BackendMessage):
    msg_type = BackendMessageType.ERROR_RESPONSE

    def __init__(self, fields: dict[str, str]) -> None:
        self.fields = fields
        self.severity = fields.get("S", "ERROR")
        self.code = fields.get("C", "")
        self.message = fields.get("M", "Unknown error")


class NoticeResponse(BackendMessage):
    msg_type = BackendMessageType.NOTICE_RESPONSE

    def __init__(self, fields: dict[str, str]) -> None:
        self.fields = fields


class EmptyQueryResponse(BackendMessage):
    msg_type = BackendMessageType.EMPTY_QUERY_RESPONSE


class ParseComplete(BackendMessage):
    msg_type = BackendMessageType.PARSE_COMPLETE


class BindComplete(BackendMessage):
    msg_type = BackendMessageType.BIND_COMPLETE


class NoData(BackendMessage):
    msg_type = BackendMessageType.NO_DATA


# ---------------------------------------------------------------------------
# Frontend message builders (client -> server)
# ---------------------------------------------------------------------------


def build_startup_message(user: str, database: str) -> bytes:
    """Build the initial startup message with protocol version and params."""
    params = f"user\x00{user}\x00database\x00{database}\x00\x00"
    params_bytes = params.encode("utf-8")
    length = 4 + 4 + len(params_bytes)  # length + version + params
    return struct.pack("!II", length, PROTOCOL_VERSION) + params_bytes


def build_password_message(password: str) -> bytes:
    """Build a cleartext password authentication message."""
    pw_bytes = password.encode("utf-8") + b"\x00"
    length = 4 + len(pw_bytes)
    return b"p" + struct.pack("!I", length) + pw_bytes


def build_md5_password_message(user: str, password: str, salt: bytes) -> bytes:
    """Build an MD5 password authentication message."""
    inner = hashlib.md5((password + user).encode("utf-8")).hexdigest()
    outer = "md5" + hashlib.md5(inner.encode("utf-8") + salt).hexdigest()
    pw_bytes = outer.encode("utf-8") + b"\x00"
    length = 4 + len(pw_bytes)
    return b"p" + struct.pack("!I", length) + pw_bytes


def build_query_message(sql: str) -> bytes:
    """Build a simple query message."""
    sql_bytes = sql.encode("utf-8") + b"\x00"
    length = 4 + len(sql_bytes)
    return b"Q" + struct.pack("!I", length) + sql_bytes


def build_parse_message(sql: str, statement_name: str = "") -> bytes:
    """Build a Parse message for extended query protocol."""
    name_bytes = statement_name.encode("utf-8") + b"\x00"
    sql_bytes = sql.encode("utf-8") + b"\x00"
    param_types = struct.pack("!H", 0)  # 0 parameter type OIDs
    length = 4 + len(name_bytes) + len(sql_bytes) + len(param_types)
    return b"P" + struct.pack("!I", length) + name_bytes + sql_bytes + param_types


def build_bind_message(
    values: list[Any],
    portal_name: str = "",
    statement_name: str = "",
) -> bytes:
    """Build a Bind message for extended query protocol."""
    portal_bytes = portal_name.encode("utf-8") + b"\x00"
    stmt_bytes = statement_name.encode("utf-8") + b"\x00"

    # Format codes: 0 = all text
    format_codes = struct.pack("!H", 0)

    # Parameters
    param_count = struct.pack("!H", len(values))
    param_data = b""
    for val in values:
        if val is None:
            param_data += struct.pack("!i", -1)  # NULL
        else:
            val_bytes = str(val).encode("utf-8")
            param_data += struct.pack("!I", len(val_bytes)) + val_bytes

    # Result format codes: 0 = all text
    result_format = struct.pack("!H", 0)

    payload = (
        portal_bytes
        + stmt_bytes
        + format_codes
        + param_count
        + param_data
        + result_format
    )
    length = 4 + len(payload)
    return b"B" + struct.pack("!I", length) + payload


def build_describe_message(
    target_type: str = "P", name: str = ""
) -> bytes:
    """Build a Describe message. target_type: 'S' for statement, 'P' for portal."""
    name_bytes = name.encode("utf-8") + b"\x00"
    length = 4 + 1 + len(name_bytes)
    return b"D" + struct.pack("!I", length) + target_type.encode("utf-8") + name_bytes


def build_execute_message(portal_name: str = "", max_rows: int = 0) -> bytes:
    """Build an Execute message."""
    portal_bytes = portal_name.encode("utf-8") + b"\x00"
    length = 4 + len(portal_bytes) + 4
    return b"E" + struct.pack("!I", length) + portal_bytes + struct.pack("!I", max_rows)


def build_sync_message() -> bytes:
    """Build a Sync message."""
    return b"S" + struct.pack("!I", 4)


def build_terminate_message() -> bytes:
    """Build a Terminate message."""
    return b"X" + struct.pack("!I", 4)


def build_sasl_initial_response(mechanism: str, client_first_message: bytes) -> bytes:
    """Build a SASLInitialResponse message."""
    mech_bytes = mechanism.encode("utf-8") + b"\x00"
    length = 4 + len(mech_bytes) + 4 + len(client_first_message)
    return (
        b"p"
        + struct.pack("!I", length)
        + mech_bytes
        + struct.pack("!I", len(client_first_message))
        + client_first_message
    )


def build_sasl_response(client_final_message: bytes) -> bytes:
    """Build a SASLResponse message."""
    length = 4 + len(client_final_message)
    return b"p" + struct.pack("!I", length) + client_final_message


# ---------------------------------------------------------------------------
# Backend message parser
# ---------------------------------------------------------------------------


def _parse_error_notice_fields(payload: bytes) -> dict[str, str]:
    """Parse error/notice response field bytes into a dict."""
    fields: dict[str, str] = {}
    pos = 0
    while pos < len(payload):
        field_type = chr(payload[pos])
        pos += 1
        if field_type == "\x00":
            break
        end = payload.index(b"\x00", pos)
        fields[field_type] = payload[pos:end].decode("utf-8")
        pos = end + 1
    return fields


def parse_backend_message(
    msg_type: int, payload: bytes
) -> BackendMessage:
    """Parse a backend message from its type byte and payload."""
    if msg_type == BackendMessageType.AUTHENTICATION:
        auth_type = struct.unpack("!I", payload[:4])[0]
        if auth_type == 0:
            return AuthenticationOk()
        elif auth_type == 3:
            return AuthenticationCleartextPassword()
        elif auth_type == 5:
            salt = payload[4:8]
            return AuthenticationMD5Password(salt)
        elif auth_type == 10:
            # SASL - parse mechanism names
            mechanisms: list[str] = []
            pos = 4
            while pos < len(payload):
                end = payload.index(b"\x00", pos)
                name = payload[pos:end].decode("utf-8")
                pos = end + 1
                if not name:
                    break
                mechanisms.append(name)
            return AuthenticationSASL(mechanisms)
        elif auth_type == 11:
            return AuthenticationSASLContinue(payload[4:])
        elif auth_type == 12:
            return AuthenticationSASLFinal(payload[4:])
        else:
            raise ValueError(f"Unknown auth type: {auth_type}")

    elif msg_type == BackendMessageType.PARAMETER_STATUS:
        sep = payload.index(b"\x00")
        name = payload[:sep].decode("utf-8")
        value = payload[sep + 1 : -1].decode("utf-8")  # strip trailing null
        return ParameterStatus(name, value)

    elif msg_type == BackendMessageType.BACKEND_KEY_DATA:
        pid, key = struct.unpack("!II", payload[:8])
        return BackendKeyData(pid, key)

    elif msg_type == BackendMessageType.READY_FOR_QUERY:
        status = chr(payload[0])
        return ReadyForQuery(status)

    elif msg_type == BackendMessageType.ROW_DESCRIPTION:
        field_count = struct.unpack("!H", payload[:2])[0]
        fields: list[FieldDescription] = []
        pos = 2
        for _ in range(field_count):
            end = payload.index(b"\x00", pos)
            name = payload[pos:end].decode("utf-8")
            pos = end + 1
            (table_oid, col_idx, type_oid, type_size, type_mod, fmt) = struct.unpack(
                "!IhIhih", payload[pos : pos + 18]
            )
            pos += 18
            fields.append(
                FieldDescription(name, table_oid, col_idx, type_oid, type_size, type_mod, fmt)
            )
        return RowDescription(fields)

    elif msg_type == BackendMessageType.DATA_ROW:
        col_count = struct.unpack("!H", payload[:2])[0]
        values: list[bytes | None] = []
        pos = 2
        for _ in range(col_count):
            length = struct.unpack("!i", payload[pos : pos + 4])[0]
            pos += 4
            if length == -1:
                values.append(None)
            else:
                values.append(payload[pos : pos + length])
                pos += length
        return DataRow(values)

    elif msg_type == BackendMessageType.COMMAND_COMPLETE:
        tag = payload[:-1].decode("utf-8")  # strip trailing null
        return CommandComplete(tag)

    elif msg_type == BackendMessageType.ERROR_RESPONSE:
        fields_dict = _parse_error_notice_fields(payload)
        return ErrorResponse(fields_dict)

    elif msg_type == BackendMessageType.NOTICE_RESPONSE:
        fields_dict = _parse_error_notice_fields(payload)
        return NoticeResponse(fields_dict)

    elif msg_type == BackendMessageType.EMPTY_QUERY_RESPONSE:
        return EmptyQueryResponse()

    elif msg_type == BackendMessageType.PARSE_COMPLETE:
        return ParseComplete()

    elif msg_type == BackendMessageType.BIND_COMPLETE:
        return BindComplete()

    elif msg_type == BackendMessageType.NO_DATA:
        return NoData()

    else:
        raise ValueError(f"Unknown backend message type: {msg_type} ({chr(msg_type)})")


# ---------------------------------------------------------------------------
# Message stream reader
# ---------------------------------------------------------------------------


class MessageReader:
    """Stateful reader that accumulates TCP data and yields complete messages."""

    def __init__(self) -> None:
        self._buffer = bytearray()

    def append(self, data: bytes) -> None:
        """Add incoming socket data to the internal buffer."""
        self._buffer.extend(data)

    def read(self) -> BackendMessage | None:
        """Read one complete backend message, or None if insufficient data."""
        if len(self._buffer) < 5:
            return None

        msg_type = self._buffer[0]
        (length,) = struct.unpack("!I", self._buffer[1:5])

        total_size = 1 + length  # type byte + length (includes itself)
        if len(self._buffer) < total_size:
            return None

        payload = bytes(self._buffer[5:total_size])
        del self._buffer[:total_size]

        return parse_backend_message(msg_type, payload)
